File: python/src/catastro_mcp/limiter.py
"""Limitador por host: techo diario acumulado (no solo ritmo) + parada por fallos consecutivos.

El aprendizaje caro: 3,2 req/s fue un ritmo razonable y aun así ~32.000 peticiones
acumuladas cortaron la IP. Lo que hay que contar es el TOTAL diario.
"""
import time
from datetime import date

from .cache import conectar

TECHO_DIARIO = 1000          # por host y día
FALLOS_MAX = 10              # fallos consecutivos → parada
PAUSA_S = 0.5                # pausa mínima entre peticiones con red


class TechoAlcanzadoError(Exception):
    pass


class Limitador:
    def __init__(self):
        self._fallos: dict[str, int] = {}
        self._ultima: dict[str, float] = {}

    def usados_hoy(self, host: str) -> int:
        con = conectar()
        try:
            row = con.execute(
                "SELECT n FROM contador WHERE host=? AND fecha=?",
                (host, date.today().isoformat()),
            ).fetchone()
            return row["n"] if row else 0
        finally:
            con.close()

    def pedir_permiso(self, host: str) -> None:
        """Llamar ANTES de cada petición. Lanza TechoAlcanzadoError si no procede."""
        if self._fallos.get(host, 0) >= FALLOS_MAX:
            raise TechoAlcanzadoError(
                f"{host}: {FALLOS_MAX} fallos consecutivos. Parada automática. "
                f"Ejecuta catastro_estado() para diagnosticar antes de reintentar."
            )
        usados = self.usados_hoy(host)
        if usados >= TECHO_DIARIO:
            raise TechoAlcanzadoError(
                f"{host}: techo diario alcanzado ({usados}/{TECHO_DIARIO}). "
                f"Se reanuda mañana. Para tandas grandes usa la descarga INSPIRE."
            )
        transcurrido = time.monotonic() - self._ultima.get(host, 0.0)
        if transcurrido < PAUSA_S:
            time.sleep(PAUSA_S - transcurrido)
        self._ultima[host] = time.monotonic()
        con = conectar()
        try:
            con.execute(
                "INSERT INTO contador(host, fecha, n) VALUES(?,?,1) "
                "ON CONFLICT(host, fecha) DO UPDATE SET n = n + 1",
                (host, date.today().isoformat()),
            )
            con.commit()
        finally:
            con.close()

    def registrar_exito(self, host: str) -> None:
        self._fallos[host] = 0

    def registrar_fallo(self, host: str) -> None:
        self._fallos[host] = self._fallos.get(host, 0) + 1


LIMITADOR = Limitador()

File: src/limiter.ts
// Limitador por host: techo diario ACUMULADO (no solo ritmo) + parada por fallos seguidos.
// El aprendizaje caro: 3,2 req/s era razonable y aun así ~32.000 acumuladas cortaron la IP.
import { conectar } from "./cache.js";

export const TECHO_DIARIO = 1000;
const FALLOS_MAX = 10;
const PAUSA_MS = 500;

const fallos = new Map<string, number>();
const ultima = new Map<string, number>();

export class TechoAlcanzadoError extends Error {}

export function usadosHoy(host: string): number {
  const row = conectar()
    .prepare("SELECT n FROM contador WHERE host=? AND fecha=?")
    .get(host, hoy()) as { n: number } | undefined;
  return row?.n ?? 0;
}

function hoy(): string {
  return new Date().toISOString().slice(0, 10);
}

export async function pedirPermiso(host: string): Promise<void> {
  if ((fallos.get(host) ?? 0) >= FALLOS_MAX) {
    throw new TechoAlcanzadoError(
      `${host}: ${FALLOS_MAX} fallos consecutivos. Parada automática. ` +
      `Ejecuta catastro_estado antes de reintentar.`);
  }
  const usados = usadosHoy(host);
  if (usados >= TECHO_DIARIO) {
    throw new TechoAlcanzadoError(
      `${host}: techo diario alcanzado (${usados}/${TECHO_DIARIO}). ` +
      `Se reanuda mañana. Para tandas grandes usa la descarga INSPIRE.`);
  }
  const delta = Date.now() - (ultima.get(host) ?? 0);
  if (delta < PAUSA_MS) await new Promise((r) => setTimeout(r, PAUSA_MS - delta));
  ultima.set(host, Date.now());
  conectar()
    .prepare("INSERT INTO contador(host,fecha,n) VALUES(?,?,1) " +
             "ON CONFLICT(host,fecha) DO UPDATE SET n=n+1")
    .run(host, hoy());
}

export function registrarExito(host: string): void { fallos.set(host, 0); }
export function registrarFallo(host: string): void { fallos.set(host, (fallos.get(host) ?? 0) + 1); }
